### ISSUE TITLE
Skip re-requesting implicit code actions if diagnostics around the cursor have not changed

### ISSUE THREAD

**Issue by mjbvz** [microsoft/vscode#215340]
https://github.com/microsoft/vscode/blob/a0185ee7cbcceda421361a2af44989dd40916a51/src/vs/editor/contrib/codeAction/browser/codeActionModel.ts#L57

I believe the implicit code actions are re-requested whenever  diagnostics in the file change. Instead I proposed that we only re-request code actions if the diagnostics on the cursor's currently line change. If I'm one line 100 for example and the diagnostics at the very top of the file change, we should no compute code actions

This is impacting some prototype work that TS is doing around sending incremental diagnostic updates

**Comment by mjbvz** [comment]
Also I'm not 100% sure this is correct but `_onMarkerChanges` should perhaps check that the cursor is not whitespace enclosed before requesting the diagnostics. I just noticed that we do that in some cases but not others

**Comment by justschen** [comment]
gotcha! some points/questions:
- atm, code actions are retriggered any time the cursor changes as well. 
- is there a case where diagnostics change elsewhere and the current cursor and diagnostics on that line don't change and nothing is typed (so cursor auto triggers don't happen)? I don't think I know of a good example right now where this happens.
- I see the `autoTrigger` happening from `onMarkersChanged` mostly when something is typed. in these cases (ie, if I type inside a string - no new markers were made, I just added a letter, but `onMarkersChanged` fires). wondering if `onMarkersChanged` is being fired in cases where it shouldn't be?
- TLDR: cases where triggers happen: cursor change and any time anything is typed

https://github.com/microsoft/vscode/assets/54879025/ffa38b54-8d09-4613-b120-f1b72dc0eccf
